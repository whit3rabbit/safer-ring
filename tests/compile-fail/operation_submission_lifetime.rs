//! Compile-fail test for operation submission lifetime constraints.
//!
//! This test verifies that the type system prevents buffer lifetime violations
//! that could lead to use-after-free bugs.

use safer_ring::{Ring, operation::Operation};
use std::pin::Pin;

fn main() {
    let ring = Ring::new(32).unwrap();
    
    {
        let mut buffer = vec![0u8; 1024];
        let pinned_buffer = Pin::new(buffer.as_mut_slice());
        
        let operation = Operation::read()
            .fd(0)
            .buffer(pinned_buffer);
        
        let _submitted = ring.submit(operation).unwrap();
        //~^ ERROR: `buffer` does not live long enough
    } // buffer is dropped here but operation is still tracked by ring
    
    // Ring continues to exist, but buffer is gone - this should not compile
}