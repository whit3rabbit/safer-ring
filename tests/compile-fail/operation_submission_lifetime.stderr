error[E0597]: `buffer` does not live long enough
  --> tests/compile-fail/operation_submission_lifetime.rs:14:38
   |
13 |         let mut buffer = vec![0u8; 1024];
   |             ---------- binding `buffer` declared here
14 |         let pinned_buffer = Pin::new(buffer.as_mut_slice());
   |                                      ^^^^^^ borrowed value does not live long enough
...
22 |     } // buffer is dropped here but operation is still tracked by ring
   |     - `buffer` dropped here while still borrowed
...
25 | }
   | - borrow might be used here, when `ring` is dropped and runs the `Drop` code for type `Ring`
   |
   = note: values in a scope are dropped in the opposite order they are defined

error[E0596]: cannot borrow `ring` as mutable, as it is not declared as mutable
  --> tests/compile-fail/operation_submission_lifetime.rs:20:26
   |
20 |         let _submitted = ring.submit(operation).unwrap();
   |                          ^^^^ cannot borrow as mutable
   |
help: consider changing this to be mutable
   |
10 |     let mut ring = Ring::new(32).unwrap();
   |         +++
